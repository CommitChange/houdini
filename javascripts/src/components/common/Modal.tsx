// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer } from 'mobx-react';
import AriaModal = require('react-aria-modal');
import { VelocityTransitionGroup } from 'velocity-react';
import 'velocity-animate';
import 'velocity-animate/velocity.ui';

export interface ModalProps {
  onClose?: () => void // if you want your modal to close, this needs to set modalActive to false
  modalActive?: boolean
  titleText?: string
  focusDialog?: boolean
  dialogStyle?: any
  childGenerator: () => any
}

class Modal extends React.Component<ModalProps, {}> {

  static defaultProps = {
    dialogStyle: { minWidth: '768px' }
  }

  render() {

    const innerModal = this.props.modalActive ? <AriaModal mounted={this.props.modalActive} titleText={this.props.titleText} focusDialog={this.props.focusDialog}
      onExit={this.props.onClose} dialogStyle={this.props.dialogStyle}>
      <header className='modal-header'>
        <h4 className='modal-header-title'>{this.props.titleText}</h4>
      </header>
      <div className="modal-body">
        {this.props.childGenerator()}
      </div>
    </AriaModal> : false

    const modal =
      <VelocityTransitionGroup
        enter={
          {
            animation: 'fadeIn',
            /* These styles are needed because, for some reason, we're 
            not able to cover the sidebar otherwise. Why? *shrug*
            */
            style: {
              position: 'fixed',
              top: '0px',
              left: '0px',
              zIndex: '5000'
            }
          }
        }

        leave={
          {
            animation: 'fadeOut',
            style: {
              position: 'fixed',
              top: '0px',
              left: '0px',
              zIndex: '5000'
            }
          }
        }
        
        runOnMount={true}>
        {innerModal}
      </VelocityTransitionGroup>;

    return modal
  }

}

export default observer(Modal)



