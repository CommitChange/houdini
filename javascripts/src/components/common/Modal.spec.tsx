// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';
import Modal, {ModalProps} from './Modal'
import {shallow} from "enzyme";
import {toJS} from "mobx";
import toJson from "enzyme-to-json";

describe('Modal', () => {
  test('nothing displayed if inactive', () => {
    let modal = shallow(<Modal childGenerator={() => <div/>}/>)

    expect(toJson(modal)).toMatchSnapshot()
  })

  test('active modal displays', () => {
    let onCloseWasCalled = false
    let modal = shallow(<Modal titleText={"title text"}
                               focusDialog={true}
                               modalActive={true}
                               onClose={() => { onCloseWasCalled = true}}
                               childGenerator={() => <div/>}/>)
    expect(toJson(modal)).toMatchSnapshot()
    let modalComponent = modal.instance() as React.Component<ModalProps, {}> //casting to modal didn't work for reasons?
    modalComponent.props.onClose()
    expect(onCloseWasCalled).toBeTruthy()
  })
})