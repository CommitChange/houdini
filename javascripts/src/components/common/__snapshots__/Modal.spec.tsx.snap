// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal active modal displays 1`] = `
<VelocityTransitionGroup
  enter={
    Object {
      "animation": "fadeIn",
      "style": Object {
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "zIndex": "5000",
      },
    }
  }
  enterHideStyle={
    Object {
      "display": "none",
    }
  }
  enterShowStyle={
    Object {
      "display": "",
    }
  }
  leave={
    Object {
      "animation": "fadeOut",
      "style": Object {
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "zIndex": "5000",
      },
    }
  }
  runOnMount={true}
>
  <Displaced
    dialogStyle={
      Object {
        "minWidth": "768px",
      }
    }
    focusDialog={true}
    mounted={true}
    onExit={[Function]}
    titleText="title text"
  >
    <header
      className="modal-header"
    >
      <h4
        className="modal-header-title"
      >
        title text
      </h4>
    </header>
    <div
      className="modal-body"
    >
      <div />
    </div>
  </Displaced>
</VelocityTransitionGroup>
`;

exports[`Modal nothing displayed if inactive 1`] = `
<VelocityTransitionGroup
  enter={
    Object {
      "animation": "fadeIn",
      "style": Object {
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "zIndex": "5000",
      },
    }
  }
  enterHideStyle={
    Object {
      "display": "none",
    }
  }
  enterShowStyle={
    Object {
      "display": "",
    }
  }
  leave={
    Object {
      "animation": "fadeOut",
      "style": Object {
        "left": "0px",
        "position": "fixed",
        "top": "0px",
        "zIndex": "5000",
      },
    }
  }
  runOnMount={true}
/>
`;
